compiler/rustc_lint/src/non_local_def.rs

```diff
@@ -215,17 +191,77 @@ impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
                     None
                 };
 
+                let mut collector = PathCollector { paths: Vec::new() };
+                collector.visit_ty(&impl_.self_ty);
+                if let Some(of_trait) = &impl_.of_trait {
+                    collector.visit_trait_ref(of_trait);
+                }
+                collector.visit_generics(&impl_.generics);
+
+                let may_move: Vec<Span> = collector
+                    .paths
+                    .into_iter()
+                    .filter_map(|path| {
+                        if path_has_local_parent(&path, cx, parent, parent_parent) {
+                            Some(path_span_without_args(&path))
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
+
+                let const_anon = matches!(parent_def_kind, DefKind::Const | DefKind::Static { .. })
+                    .then_some(span_for_const_anon_suggestion);
+
+                let may_remove = match &impl_.self_ty.kind {
+                    TyKind::Ptr(mut_ty) | TyKind::Ref(_, mut_ty)
+                        if ty_has_local_parent(&mut_ty.ty.kind, cx, parent, parent_parent) =>
+                    {
+                        let type_ =
+                            if matches!(impl_.self_ty.kind, TyKind::Ptr(_)) { "*" } else { "&" };
+                        let part = format!("{}{}", type_, mut_ty.mutbl.prefix_str());
+                        Some((impl_.self_ty.span.shrink_to_lo().until(mut_ty.ty.span), part))
+                    }
+                    _ => None,
+                };
+
+                let impl_span = item.span.shrink_to_lo().to(impl_.self_ty.span);
+                let mut ms = MultiSpan::from_span(impl_span);
+
+                let (self_ty_span, self_ty_str) =
+                    self_ty_kind_for_diagnostic(&impl_.self_ty, cx.tcx);
```

Comment:
Given that this is only used to get the type name for the type and trait in the `impl` block, I would avoid it: we're already pointing at the appropriate spans, so naming them is redundant with just the line above. Instead you can just leave it as "this is not local" and that's it. The only case where naming the type is useful is when it ends up in a message of a note or help, on its own and will otherwise not be seen in context (which can also happen with rust-analyzer).

This change would also preclude the need of depending on hir_pretty, which is an added bonus.

Reply:
I wanted to mentioned them for the macro generated case since otherwise the span is just pointing to the macro it-self which isn't helpful.